kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -123,11 +123,14 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
     }
 
     lock(owner)
-    try {
-        return action()
-    } finally {
+    val result = try {
+        action()
+    } catch (e: Throwable) {
         unlock(owner)
+        throw e
     }
+    unlock(owner)
+    return result
```

Comment:
I'm sure you'll agree that this is an unnatural way to write this. You shouldn't trust me to remember not to "fix" the style of code accidentally unless you drop a comment!

Reply:
I agree, that's why I added the test. I'm not sure at all what to write in the comment other than "see test […] for an explanation of why it's an issue"

Reply:
Added a comment.